Cleanup of circuit library ``control`` methods

alexanderivrii opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary (updated on 19.11.2025)

This PR cleans up and unifies the handling of ``gate.control`` methods for gates in our circuit library. The principles behind these changes are:

* using native gate classes for controlled gates when available
* set the default value for ``annotated`` to ``None`` for all ``gate.control`` methods. This change has no effect on the synthesized circuits since ``None`` is essentially treated as ``False``.
* improve type hints and docstrings for ``gate.control`` methods across the library

These changes do not affect the transpiled circuits (or only make the results better).



### Details and comments

The method `gate.control(annotated=True)` no longer returns an annotated operation when a native controlled-gate class is available. This change enables the most efficient decompositions of such gates, and is consistent with how the argument ``annotated`` is used across the standard circuit library. The affected gates are:
    * Single-controlled ``H``, ``S``, ``Sdg``, ``U3``, ``Y``, ``Z``, ``SX``, ``RX``, ``RY``, ``RZ``, ``Swap`` and ``CZ`` gates;
    * Double-controlled ``Z``-gate;
    * arbitrarily-controlled ``Phase``, ``CPhase``, ``MCPhase``, ``U1``, ``CU1``, ``MCU1`` and ``MCMT`` gates.

The default value for ``annotated`` in ``QuantumCircuit.control`` was also changed from ``False`` to ``None``. This has absolutely no effiect on the sythesized circuits, but again makes the default argument consistent with the rest of the control methods.

Most importantly, this brings us significantly closer to set ``annotated=True`` by default in Qiskit 3.0, allowing significantly more efficient decompositions of nested circuits with controlled gates. 

Some additional history: see #12752, #13507.

alexanderivrii committed:
fix merge conflicts and write release note

alexanderivrii committed:
lint

ShellyGarion reviewed (commented):
As far as I understand, we only need a `control` method for gates that have a specific controlled gate (H, P, RX, RY, RZ, S, Sdg, SX, SXdg, U, U1, U3, X, Y, Z, CX, CZ, CP, SWAP, C3X, C4X, CCX, RCCX).

Follwoing #13507 - is there a reason to have a `control` method for the 2-qubit parametrized gates? (RZZ, RXX, RYY, RZX, XX+YY, XX-YY)

ShellyGarion commented on test/python/circuit/test_controlled_gate.py:
```diff
@@ -1772,121 +1798,46 @@ def test_control_label_1(self, gate, args):
+        (CXGate(), 2),
+    )
+    @unpack
+    def test_native_controlled_gates(self, base_gate, num_controls):
```
these tests are for gates that have multi-controlled version? perhaps it's better to say "multi".
also, I think you should add here:
CCX, C3X, C4X

alexanderivrii replied:
Terminology-wise: a _controlled_ gate is a a gate with an _arbitrary number_ of controls; we already write `gate.control(...)` and `ControlledGate(...)` to refer to such a gate.

I can add more tests if you think we really need these; this was not supposed to be fully exhaustive but rather a sanity check that `gate.control(annotated=...)` behaves as we expect it to.

ShellyGarion commented on test/python/circuit/test_controlled_gate.py:
```diff
@@ -1772,121 +1798,46 @@ def test_control_label_1(self, gate, args):
+        (SwapGate(), 2),
+        (U3Gate(0.1, 0.2, 0.3), 2),
+    )
+    @unpack
```
these gates should have only a controlled version but not multi-controlled?
I think you should add here:
S, Sdg, RCCX

ShellyGarion replied:
actually, I'm not sure about RCCX. is C(RCCX) = RC3X?

alexanderivrii replied:
Same here, I can add more tests if you think we need them, but the tests are not supposed to be fully exhaustive, but rather a sanity check that gate.control(annotated=...) behaves as we expect it to. 

Indeed, `C(RCCX)` is not `RC3X` (or even if they magically happen to be the same, there is no reason they need to be).

ShellyGarion reviewed (commented):
A few minor comments about the tests. I think that a few cases might be still missing.
btw, I'm not sure if we have tests for `MCMTGate.control()`.

alexanderivrii committed:
removing unnecessary explicit control methods for RXX, RZZ, etc.

We have added control methods in #12752 as the only way to support controlled
versions of these gates with symbolic parameters, however after #13507 these
control methods are no longer required.

alexanderivrii committed:
also simplifying the special treatment of rx, ry, etc. gates

alexanderivrii committed:
update release notes

alexanderivrii commented:
> Follwoing https://github.com/Qiskit/qiskit/pull/13507 - is there a reason to have a control method for the 2-qubit parametrized gates? (RZZ, RXX, RYY, RZX, XX+YY, XX-YY)

Amazing observation, Shelly! Indeed we can now completely remove the special handling of controlled pararametized gates that were added as a part of #12752, leading to a significantly simpler and cleaner code. I have done this change, and updated both the PR description and the release notes accordingly.

So, what we should have right now is as follows:
* If there is a native gate class representing `gate.control()`, we use that as a controlled gate. In this case the argument `annotated` is completely irrelevant, the controlled gate is neither a `ControlledGate` nor an `AnnotatedOperation`.
* If not, and if `annotated=False`, we create a `ControlledGate`, #13507 makes sure that these gates have a valid definition circuit (possibly depending on free parameters).
* If not, and if `annotated=True`, we create an `AnnotatedOperation`, #12752 ensures that the parameters of such an annotated operations are treated correctly.

The default value of `annotated` is `None` which of right now is completely identical to `False`. However, later we want to deprecate the value of `None` and let the default be `True` instead. This change would allow performing more complex optimizations, and delegate synthesis to transpiler from the gate creation time.

alexanderivrii committed:
pylint

alexanderivrii commented:
For MCMT gates: we have the basic correctness checks in `test_controlled_gate`, where we check that the controlled versions of MCMT gates have expected matrices.

ShellyGarion commented:
Another question about the tests, here we check `annotated=[False, None]`, should we also check `annotated=True` ?

https://github.com/Qiskit/qiskit/blob/96d29c7c6d703294c4d91b52edef7aa62cd8cbe9/test/python/circuit/test_controlled_gate.py#L1488

ShellyGarion commented:
The API docs look great - thanks for the effort :)

alexanderivrii committed:
revamping an older test

alexanderivrii commented:
> Another question about the tests, here we check annotated=[False, None], should we also check annotated=True ?

Done in [cb49035](https://github.com/Qiskit/qiskit/pull/15209/commits/cb49035cf7c8231844d10c01abf4a24fcf8d816e). This is indeed an old test that could have been improved.

ShellyGarion reviewed (approved):
@alexanderivrii - thanks for the effort! the control methods look much better now!
